sound/soc/sof/intel/hda-dai.c

```diff
@@ -319,23 +302,37 @@ static int hda_dai_suspend(struct hdac_bus *bus)
 		 * explicitly during suspend.
 		 */
 		if (hext_stream->link_substream) {
-			struct snd_soc_dai *cpu_dai;
+			const struct snd_sof_dai_ops *ops;
+			struct snd_sof_widget *swidget;
+			struct snd_soc_dapm_widget *w;
 			struct snd_soc_dai *codec_dai;
+			struct snd_soc_dai *cpu_dai;
+			struct snd_sof_dev *sdev;
+			struct snd_sof_dai *sdai;
 
 			rtd = asoc_substream_to_rtd(hext_stream->link_substream);
 			cpu_dai = asoc_rtd_to_cpu(rtd, 0);
 			codec_dai = asoc_rtd_to_codec(rtd, 0);
+			w = snd_soc_dai_get_widget(cpu_dai, hdac_stream(hext_stream)->direction);
+			swidget = w->dobj.private;
+			sdev = snd_soc_component_get_drvdata(swidget->scomp);
+			sdai = swidget->private;
+			ops = sdai->ops;
 
 			ret = hda_link_dma_cleanup(hext_stream->link_substream,
 						   hext_stream,
 						   cpu_dai, codec_dai);
 			if (ret < 0)
 				return ret;
 
-			/* for consistency with TRIGGER_SUSPEND we free DAI resources */
-			ret = hda_dai_hw_free_ipc(hdac_stream(hext_stream)->direction, cpu_dai);
-			if (ret < 0)
-				return ret;
+			/* for consistency with TRIGGER_SUSPEND  */
+			if (ops && ops->dma_ops && ops->dma_ops->post_trigger) {
+				ret = ops->dma_ops->post_trigger(sdev, cpu_dai,
+								 hext_stream->link_substream,
+								 SNDRV_PCM_TRIGGER_SUSPEND);
```

Comment:
I didn't really follow how the hw_free and post-trigger were related. The naming points to different phases in the ALSA world.